### ISSUE TITLE
A lot of the measured interface-call cost is GC

### ISSUE THREAD

**Issue by twotwotwo** [teh-cmc/interface-bench#1]
It turns out a lot of the time the interface benchmark was taking was in GC related to the particular way interfaces are being used here, not cost that's inherent to all interface calls. 

As background--since Go 1.4, when you wrap a non-pointer type like a uint64 in an interface, the interface value holds a pointer, and the actual int is allocated elsewhere. You can see the allocations by using MemStats with the original code: https://play.golang.org/p/UUlVZp7x-r. 

If that cost were present every time someone makes an interface call (or most of the time) you'd want to count it, of course, but it often isn't.

You don't get an extra allocation when you store a pointer type in an interface var, which is probably the most common case, like when you put a `*bufio.Writer` or `*os.File` in an `io.Writer`. 

You also get only one extra allocation if you store a non-pointer type in an interface var once, then make lots of calls on it, like when you wrap a large slice in a `sort.Interface` and call `sort.Sort` on it. 

After this benchmark got posted on Twitter and I noticed it was allocating for each iface call, I rigged up a variation where you call Add on an `*Int`, so you don't get an alloc per call: https://play.golang.org/p/OgAcagFmZi

The results I got were 0 bytes allocated and 177.376219ms for the concrete calls, 16 bytes allocated and 210.502363ms for the indirect calls. I'm guessing the switch to using pointers was what made the concrete calls slower. Compiling with -gcflags -m shows the concrete call to Add still gets inlined.

If it works for you, I can submit that code and the updates that go with it in a PR so anyone stumbling on this in the future can see the interface call costs without the GC situation. Or if you want to integrate the code yourself, or integrate the info here in some other way, either sounds great to me.

**Comment by teh-cmc** [comment]
Thanks for the detailed explanation!

I had totally forgotten about the changes regarding interfaces & one-word scalar values that shipped with the 1.4 release.  
Without a doubt, my benchmark is completely wrong as it stands.

In light of this, my first thought was that all these implicit allocations were causing stop-the-world GC pauses, hence driving latency through the roof; so I tried disabling the garbage collector while running the original code. It turns out the slowdown is not GC-related per-se, the results are the same even with the GC completely disabled:  

```
$ GOGC=off go run main.go
[concrete]  computed 100000000 sums in 42.698356ms
[interface] computed 100000000 sums in 2.99054357s
```

Running a pprof trace shows that the actual bottleneck is, quite simply, the overhead of allocating and copying around 100 million integers as part of the process of converting types to interfaces (`runtime.convT2I`):  

```
  flat  flat%   sum%        cum   cum%
 740ms 28.24% 28.24%     1180ms 45.04%  runtime.mallocgc
 340ms 12.98% 41.22%     1520ms 58.02%  runtime.newobject
 270ms 10.31% 51.53%      270ms 10.31%  runtime.mach_semaphore_signal
 260ms  9.92% 61.45%     2490ms 95.04%  main.main
 220ms  8.40% 69.85%     2090ms 79.77%  runtime.convT2I
 180ms  6.87% 76.72%      180ms  6.87%  runtime.memmove
 150ms  5.73% 82.44%      330ms 12.60%  runtime.typedmemmove
 130ms  4.96% 87.40%      130ms  4.96%  main.(*Int).Sum
 100ms  3.82% 91.22%      100ms  3.82%  runtime.prefetchnta
  80ms  3.05% 94.27%       80ms  3.05%  runtime.(*mspan).sweep.func1
```

In any case, your solution still applies: storing a pointer in the interface and using in-place updates instead of return values looks like the only way to avoid serious performance issues here.

I've updated the README to add a detailed explanation for all of this (I like to keep notes :) ), as well as a mention to your ticket, of course.

Thanks!

**Comment by twotwotwo** [comment]
Oh, interesting that it's the same when GC is off! Props again for all of the digging here.
